VOCSegmentation, VOCDetection, linting passing, examples.

bpinaya opened the pull request:
Hey there guys, I hope this PR can revive some of the efforts that past PRs implemented, I'm referring to:
- #86 by @ellisbrown
- #37 by @desimone

This PR supports VOCSegmentation and VOCDetection for multiple years (2007 to 2012). 2006 and 2005 are not supported due to the very different format.
I tried to address many of the comments of the pasts two PRs. For the VOCDetection part `__getitem__` will return the image and a list of bounding boxes of the format `[[xmin, ymin, xmax, ymax, ind],[...],...]` as PR 37 directly instead of a ET Element.
I think VOC is a very iconic dataset for both detection and segmentation so it'd be very useful to have it ready to use on pytorch/vision.

If there are any comments or suggestions let me know, I can put some time on implementing recommendations or so.
There are two examples (jupyter notebook gist) of the dataset being loaded:
- This one is for [segmentation](https://gist.github.com/bpinaya/3017768b87d93153c42547b84ce052f3)
- This one is for [detection](https://gist.github.com/bpinaya/c39f9f6adf7f4110e3c121155c3d371b)

I've tested for all the years and regarding the transforms, I'm using the same structure other datasets used, I've seen implementations where a `joint_transforms` is passed as argument, and that transform is applied to both input and target, I guess that'd work for segmentation but not much for detection so that's why it's not used.

adamjstewart commented on torchvision/datasets/voc.py:
```diff
@@ -0,0 +1,264 @@
+    Args:
+        root (string): Root directory of the VOC Dataset.
+        year (string, optional): The dataset year, supports years 2007 to 2012.
+        image_set (string, optional): Select the image_set to use, ``train, trainval or val``
```
I would wrap `train`, `trainval`, and `val` in double backticks separately.

bpinaya replied:
Good catch! And it'd be helpful if you tried some years (2011,2018) I did the test for all but just in case.

adamjstewart commented on torchvision/datasets/voc.py:
```diff
@@ -0,0 +1,264 @@
+                relative coordinates like``[[xmin, ymin, xmax, ymax, ind], [...], ...]``.
+        """
+        _img = Image.open(self.images[index]).convert('RGB')
+        _target = self._get_bboxes(ET.parse(self.annotations[index]).getroot())
```
Out of curiosity, why do all of your local variables begin with an underscore? I've never seen that style in Python code before.

desimone replied:
https://www.python.org/dev/peps/pep-0008/#descriptive-naming-styles
http://google.github.io/styleguide/pyguide.html#3164-guidelines-derived-from-guidos-recommendations

adamjstewart replied:
I agree that the leading underscore makes sense for internal functions/attributes or global variables that you don't want imported with `from foo import *`, but it doesn't really make sense inside a function where the scope already prevents them from being accessed externally. It's just redundant.

fmassa replied:
Can we instead return the raw result from `ET.parse().getroot()`? Or maybe a dict with the full parsing of the xml?
It's up to the user to decide what they actually want for their target, and we are forcing them to use one format now (which doesn't hold all the information from the dataset, such as truncated / occluded / etc).

fmassa commented:
Hi,

I'm just about to get https://github.com/facebookresearch/maskrcnn-benchmark/pull/131 merged, and I'll be pushing a more genera version of this here. I can coordinate with you guys so that we can get this finally merged.

bpinaya commented:
Hi @fmassa, just checked the PR, seems really robust. Keep me in the loop if you need help with testing or implementation.

fmassa commented:
I've just finished testing it and got some reasonable results with the codebase. I'll look into merging the PR todayish.

bpinaya commented:
Looking forwards to it!

ellisbrown commented on torchvision/datasets/voc.py:
```diff
@@ -0,0 +1,264 @@
+    'car', 'cat', 'chair', 'cow', 'diningtable', 'dog', 'horse', 'motorbike',
+    'person', 'pottedplant', 'sheep', 'sofa', 'train', 'tvmonitor'
+]
+DATASET_YEAR_DICT = {
```
Why not make this a dict of dicts instead of forcing the user to remember the order of elements in the list?

```suggestion
DATASET_YEAR_DICT = {
    '2012': {
        'url': 'http://host.robots.ox.ac.uk/pascal/VOC/voc2012/VOCtrainval_11-May-2012.tar',
        'filename': 'VOCtrainval_11-May-2012.tar',
        'md5': '6cd6e144f989b92b3379bac3b3de84fd',
        'base_dir': 'VOCdevkit/VOC2012'
    },
    '2011': {
        'url': 'http://host.robots.ox.ac.uk/pascal/VOC/voc2011/VOCtrainval_25-May-2011.tar',
        'filename': 'VOCtrainval_25-May-2011.tar',
        'md5': '6c3384ef61512963050cb5d687e5bf1e',
        'base_dir': 'TrainVal/VOCdevkit/VOC2011'
    },
    '2010': {
        'url': 'http://host.robots.ox.ac.uk/pascal/VOC/voc2010/VOCtrainval_03-May-2010.tar',
        'filename': 'VOCtrainval_03-May-2010.tar,
        'md5': 'da459979d0c395079b5c75ee67908abb',
        'base_dir': 'VOCdevkit/VOC2010'
    },
    '2009': {
        'url': 'http://host.robots.ox.ac.uk/pascal/VOC/voc2009/VOCtrainval_11-May-2009.tar',
        'filename': 'VOCtrainval_11-May-2009.tar'',
        'md5': '59065e4b188729180974ef6572f6a212',
        'base_dir': 'VOCdevkit/VOC2009'
    },
    '2008': {
        'url': 'http://host.robots.ox.ac.uk/pascal/VOC/voc2008/VOCtrainval_14-Jul-2008.tar',
        'filename': 'VOCtrainval_11-May-2012.tar',
        'md5': '2629fa636546599198acfcfbfcf1904a',
        'base_dir': 'VOCdevkit/VOC2008'
    },
    '2007': {
        'url': 'http://host.robots.ox.ac.uk/pascal/VOC/voc2007/VOCtrainval_06-Nov-2007.tar',
        'filename': 'VOCtrainval_06-Nov-2007.tar',
        'md5': 'c52e279531787c972589f7e41ab4ae64',
        'base_dir': 'VOCdevkit/VOC2007'
    }
}
```

ellisbrown commented on torchvision/datasets/voc.py:
```diff
@@ -0,0 +1,264 @@
+        self.transform = transform
+        self.target_transform = target_transform
+        self.image_set = image_set
+        _base_dir = DATASET_YEAR_DICT[year][3]
```
```suggestion
        _base_dir = DATASET_YEAR_DICT[year]['base_dir']
```

ellisbrown commented on torchvision/datasets/voc.py:
```diff
@@ -0,0 +1,264 @@
+        self.year = year
+        self.url = DATASET_YEAR_DICT[year][0]
+        self.filename = DATASET_YEAR_DICT[year][1]
+        self.md5 = DATASET_YEAR_DICT[year][2]
```
```suggestion
        self.md5 = DATASET_YEAR_DICT[year]['md5']
```

ellisbrown commented on torchvision/datasets/voc.py:
```diff
@@ -0,0 +1,264 @@
+        self.root = root
+        self.year = year
+        self.url = DATASET_YEAR_DICT[year][0]
+        self.filename = DATASET_YEAR_DICT[year][1]
```
```suggestion
        self.filename = DATASET_YEAR_DICT[year]['filename']
```

ellisbrown commented on torchvision/datasets/voc.py:
```diff
@@ -0,0 +1,264 @@
+                 target_transform=None):
+        self.root = root
+        self.year = year
+        self.url = DATASET_YEAR_DICT[year][0]
```
```suggestion
        self.url = DATASET_YEAR_DICT[year]['url']
```

ellisbrown commented on torchvision/datasets/voc.py:
```diff
@@ -0,0 +1,264 @@
+                 target_transform=None):
+        self.root = root
+        self.year = year
+        self.url = DATASET_YEAR_DICT[year][0]
```
```suggestion
        self.url = DATASET_YEAR_DICT[year]['url']
```

ellisbrown commented on torchvision/datasets/voc.py:
```diff
@@ -0,0 +1,264 @@
+        self.root = root
+        self.year = year
+        self.url = DATASET_YEAR_DICT[year][0]
+        self.filename = DATASET_YEAR_DICT[year][1]
```
```suggestion
        self.filename = DATASET_YEAR_DICT[year]['filename']
```

ellisbrown commented on torchvision/datasets/voc.py:
```diff
@@ -0,0 +1,264 @@
+        self.year = year
+        self.url = DATASET_YEAR_DICT[year][0]
+        self.filename = DATASET_YEAR_DICT[year][1]
+        self.md5 = DATASET_YEAR_DICT[year][2]
```
```suggestion
        self.md5 = DATASET_YEAR_DICT[year]['md5']
```

ellisbrown commented on torchvision/datasets/voc.py:
```diff
@@ -0,0 +1,264 @@
+        self.class_to_ind = class_to_ind or dict(
+            zip(VOC_CLASSES, range(len(VOC_CLASSES))))
+        self.keep_difficult = keep_difficult
+        _base_dir = DATASET_YEAR_DICT[year][3]
```
```suggestion
        _base_dir = DATASET_YEAR_DICT[year]['base_dir']
```

bpinaya replied:
Solid suggestion @ellisbrown , I've committed those changes with this new github "apply suggestion" feature, quite handy. Thanks for the feedback!

fmassa commented on torchvision/datasets/voc.py:
```diff
@@ -0,0 +1,270 @@
+    def __len__(self):
+        return len(self.images)
+
+    def _get_bboxes(self, target):
```
I'd probably let the user write this down. Maybe what would be the most user-friendly would be to parse the ET and return a nested dict?

bpinaya replied:
Indeed, I thought about it, maybe a dict would be better, but I saw [here](https://github.com/facebookresearch/maskrcnn-benchmark/pull/207/files#diff-fd37bdf323744e6fb84a9f8d11b9e10cR76) that they pass a Bb list. I think whatever is easier for the end user maybe. Regarding the iteration of the ET, any ideas to make it recursive and elegant? I implemented a function but it's way to hacky.

fmassa replied:
In `maskrcnn-benchmark`, we have a dedicated class `BoxList` which is used everywhere in the codebase.
My original plan was to move `BoxList` to `torchvision`, but it needs to mature a bit more before we move it here.

About the ET, I suppose there is a function call that we can use that would enable us to get it recursively? Probably something like [this that I wrote for lua](https://github.com/fmassa/object-detection.torch/blob/refactoring2/DataSetPascal.lua#L167-L185)

fmassa commented on torchvision/datasets/voc.py:
```diff
@@ -0,0 +1,270 @@
+
+        _splits_dir = os.path.join(_voc_root, 'ImageSets/Main')
+
+        _split_f = os.path.join(_splits_dir, image_set.rstrip('\n') + '.txt')
```
do we need the `rstrip`? just out of curiosity

fmassa commented on torchvision/datasets/voc.py:
```diff
@@ -0,0 +1,270 @@
+                 image_set='train',
+                 download=False,
+                 class_to_ind=None,
+                 keep_difficult=False,
```
I think `keep_difficult` could be part of what the user pass to the `target_transform`.

fmassa commented on torchvision/datasets/voc.py:
```diff
@@ -0,0 +1,270 @@
+        self.images = []
+        self.annotations = []
+        with open(os.path.join(_split_f), "r") as lines:
+            for line in lines:
```
Can't we instead do
```python
with open(os.path.join(_split_f), "r") as f:
    image_names = f.readlines()
```
? I believe this strips out the `\n` in the end, and is a bit faster than the current version.

bpinaya replied:
I've checked and it doesn't strip the `\n`
the value of image_names is:
```
['2008_000008\n', '2008_000015\n', '2008_000019\n', '2008_000023\n', '2008_000028\n', '2008_000033\n', '2008_000036\n', '2008_000037\n', '2008_000041\n', '2008_000045\n', '2008_000053\n', '2008_000060\n', '2008_000066\n', '2008_000070\n', '2008_000074\n', '2008_000085\n', '2008_000089\n', '2008_000093\n', '2008_000095\n', '2008_000096\n', '2008_000097\n', '2008_000099\n', '2008_000103\n', '2008_000105\n', '2008_000109\n', '2008_000112\n'...
```

fmassa replied:
ok. I think what I had done then was something like
```
with open(os.path.join(_split_f), "r") as f:
    image_names = [x.strip() for x in f.readlines()]
```
but do it as you think it's better.

adamjstewart commented:
Can you also update `docs/source/datasets.rst` so that this new Dataset appears in the documentation?

bpinaya commented:
I think the better approach would be to pass a dict of the parsed `ET.parse().getroot()`. I'm struggling a bit to implement a recursive function to parse it correctly so if anyone has some tips, they are appreciated.

bpinaya commented:
@fmassa I followed your recommendations and added this instead for loading:
```python
        with open(os.path.join(_split_f), "r") as f:
            file_names = [x.strip() for x in f.readlines()]

        self.images = [os.path.join(_image_dir, x + ".jpg") for x in file_names]
        self.annotations = [os.path.join(_annotation_dir, x + ".xml") for x in file_names]
```
In the detection part and this in the segmentation part:
```python
        with open(os.path.join(_split_f), "r") as f:
            file_names = [x.strip() for x in f.readlines()]

        self.images = [os.path.join(_image_dir, x + ".jpg") for x in file_names]
        self.masks = [os.path.join(_mask_dir, x + ".png") for x in file_names]
```
Removed `keep_difficult` and `_get_bboxes()` and replaced it with `parse_voc_xml` that converts the xml to a dictionary like this (formatted):
```python
{
    'annotation': {
        'filename': '2010_001870.jpg',
        'folder': 'VOC2010',
        'object': [{
            'name': 'person',
            'bndbox': {
                'xmax': '325',
                'xmin': '110',
                'ymax': '375',
                'ymin': '72'
            },
            'difficult': '0',
            'occluded': '0',
            'pose': 'Frontal',
            'truncated': '1',
            'part': [{
                'name': 'head',
                'bndbox': {
                    'xmin': '180',
                    'ymin': '75',
                    'xmax': '286',
                    'ymax': '211'
                }
            }, {
                'name': 'hand',
                'bndbox': {
                    'xmin': '225',
                    'ymin': '284',
                    'xmax': '292',
                    'ymax': '342'
                }
            }]
        }, {
            'name': 'person',
            'bndbox': {
                'xmax': '490',
                'xmin': '57',
                'ymax': '327',
                'ymin': '73'
            },
            'difficult': '0',
            'occluded': '1',
            'pose': 'Frontal',
            'truncated': '1',
            'part': [{
                'name': 'head',
                'bndbox': {
                    'xmin': '283',
                    'ymin': '88',
                    'xmax': '360',
                    'ymax': '226'
                }
            }, {
                'name': 'hand',
                'bndbox': {
                    'xmin': '58',
                    'ymin': '235',
                    'xmax': '116',
                    'ymax': '329'
                }
            }, {
                'name': 'hand',
                'bndbox': {
                    'xmin': '356',
                    'ymin': '136',
                    'xmax': '413',
                    'ymax': '215'
                }
            }]
        }],
        'segmented': '0',
        'size': {
            'depth': '3',
            'height': '375',
            'width': '500'
        },
        'source': {
            'annotation': 'PASCAL VOC2010',
            'database': 'The VOC2010 Database',
            'image': 'flickr'
        }
    }
}
```
As you can see it handles `part` and `objects` ok.
The only downside would be that the user needs to pass a target transform otherwise it won't be usable directly if a `target_transform` is not passed. I can write a transform and added to transforms, or shall we leave that to the user?

fmassa commented on torchvision/datasets/voc.py:
```diff
@@ -0,0 +1,252 @@
+        self.transform = transform
+        self.target_transform = target_transform
+        self.image_set = image_set
+        self.class_to_ind = class_to_ind or dict(
```
Can we remove `class_to_ind`? It's not used anymore.

fmassa commented on torchvision/datasets/voc.py:
```diff
@@ -0,0 +1,252 @@
+    def __len__(self):
+        return len(self.images)
+
+    def parse_voc_xml(self, node):
```
This looks great, thanks!
Just to double check, did you try parsing all the images in say VOC2012? I know that some images have a single object and that might require some special handling, just want to verify that this is indeed being taken care here.

bpinaya commented:
Oh indeed I was not using `class_to_ind` anymore, removed it and removed `VOC_CLASSES` too. And I just checked with VOC2012, it seems it's working for parts that have a single object.
[This gist](https://gist.github.com/bpinaya/cda500afa3539727c49ef42b1e9126e2) is the data loading of VOC2010, for `target_transform` I just passed a function that prints it and returns a 0. 
Also tested for VOC2012 in [this gist](https://gist.github.com/bpinaya/5a1bf6e23580e84492ae4bb53c65a351), as you can see it works when the object has one or multiple bounding boxes. Picture bellow for example, but you might just check the gists.
![image](https://user-images.githubusercontent.com/2473733/49585245-1f2bb400-f95e-11e8-9df3-15c0b01a0a97.png)

fmassa commented:
@feixiangdekaka you can use `torch.utils.data.ConcatDataset` to concatenate two datasets created from different years, or even COCO + Pascal.

feixiangdekaka commented:
data_transforms = (
    transforms.Compose([
        transforms.RandomResizedCrop(224),
        transforms.RandomHorizontalFlip(),
        transforms.ToTensor(),
        transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225])
    ]))

dataset= torchvision.datasets.VOCDetection(root="~/.torch",year="2007",download=True,transform=data_transforms)
trainloader = torch.utils.data.DataLoader(dataset1, batch_size=64,
                                          shuffle=True, num_workers=2)

how to correctly iter the datesets?
(1)next(iter(trainloader))
(2)for i, images,target in enumerate(trainloader):
but get error,any hint?

adamjstewart commented:
@feixiangdekaka Two things look wrong here. (1) you define `dataset` but refer to it as `dataset1` later in the code. I'm going to assume this is a typo. (2) you need parentheses around `(images, target)` as they are returned as a single tuple. It should look like:
```python
for i, (images, target) in enumerate(trainloader):
```

feixiangdekaka commented:
when I use cifa-10 it can get right result!
when using voc  with bathsize=1  it is right,when bathsize >1 it return the error:
  
  def _process_next_batch(self, batch):
        self.rcvd_idx += 1
        self._put_indices()
        if isinstance(batch, ExceptionWrapper):
            raise batch.exc_type(batch.exc_msg)
        return batch
raise ExceptionWapper ERROR!

fmassa commented:
@feixiangdekaka can you run with `num_workers=0` so that the error message is clearer, and paste the results here?
